test/SlowTests/Issues/RavenDB-18554.cs

```diff
@@ -0,0 +1,212 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using FastTests;
+using FastTests.Client;
+using NuGet.ContentModel;
+using Orders;
+using Raven.Client.Documents;
+using Raven.Client.Documents.Indexes;
+using Raven.Client.Documents.Operations;
+using Raven.Client.Documents.Operations.Indexes;
+using Raven.Client.Documents.Session;
+using Raven.Client.Exceptions.Database;
+using Raven.Client.ServerWide;
+using Raven.Client.ServerWide.Operations;
+using Raven.Server;
+using Raven.Server.Commercial;
+using Raven.Server.Documents.Indexes;
+using Tests.Infrastructure;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace SlowTests.Issues
+{
+    public class RavenDB_18554 : ClusterTestBase
+    {
+        public RavenDB_18554(ITestOutputHelper output) : base(output)
+        {
+        }
+
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task QueriesShouldFailoverIfDatabaseIsCompacting(bool cluster = false)
+        {
+            Options storeOptions;
+            RavenServer leader = null;
+            if (cluster)
+            {
+                List<RavenServer> nodes;
+                (nodes, leader) = await CreateRaftCluster(2);
+                storeOptions = new Options {Server = leader, ReplicationFactor = nodes.Count, RunInMemory = false};
+            }
+            else
+            {
+                storeOptions = new Options {RunInMemory = false};
+            }
+            
+            using (var store = GetDocumentStore(storeOptions))
+            {
+                // Prepare Server For Test
+                string categoryId;
+                using (var session = store.OpenAsyncSession())
+                {
+                    Category c = new Category {Name = $"n0", Description = $"d0"};
+                    await session.StoreAsync(c);
+                    await session.SaveChangesAsync();
+                    categoryId = c.Id;
+                }
+                
+                var index = new Categoroies_Details();
+                index.Execute(store);
+                Indexes.WaitForIndexing(store);
+
+                string[] indexNames = { index.IndexName };
+                CompactSettings settings = new CompactSettings {DatabaseName = store.Database, Documents = true, Indexes = indexNames};
+
+                var database = cluster ? await GetDatabase(leader, store.Database) : await GetDatabase(store.Database);
+
+                // Test
+                Exception e = null;
+                database.ForTestingPurposesOnly().CompactionAfterDatabaseUnloadAction = () =>
+                {
+                    try
+                    {
+                        using (var session = store.OpenSession())
+                        {
+                            session.Query<Categoroies_Details.Entitiy, Categoroies_Details>()
```

Comment:
Check that we got the correct result here (for cluster)

Reply:
Done